## <center>Lab 2: Estimating Integrals</center>

<center><img src = "Integration.png" height = "300", width = "300"></center>

In this graded lab session, our objective is to implement **algorithms** for performing numerical integration.  You will practice:

* Writing mathematical **expressions**

* Writing and using **functions** to solve problems

* Writing and using **selection flow control** (conditional execution statements)

* Writing and using **iteration flow control** (**while**/for loops)

**Before You Start:** Make sure that you can open, run, and save a Jupyter Notebook. If you have any problems, please see the instructor.

### Task 1

Review the class materials on numerical integration.  You have been provided three handouts/supplements to support this lab:

*  **IntegrationReference.pdf:** this excerpt from *Calculus and Analytic Geometry* by Sherman Stein provides an explanation of the iterative algorithms we will use to conduct numerical integration.


* **IntegrationPseudocode.pdf:** this document provides *pseudo code* for implementing the algorithms discussed by Stein (which you should find very helpful).


* **IntegrationSolutions.pdf:**  this document provides closed form solutions to the numerical integrations you will perform  


* *NOTE:* **Lab_2_Results.doc** will serve as a worksheet to formally capture your specific results for each task stipulated in this notebook. You will submit both your modified Jupyter notebook AND the results worksheet as described at the end of this notebook.

### Task 2

Review the organization of this notebook (read the entire notebook before you do anything else). This file contains four predefined functions for  your use: 

* ```f0(x)```
* ```f1(x)``` 
* ```f2(x)```
* ```print result()```

These functions should NOT be modified. Make sure you understand what they do, and use them as provided.

#### Predefined Functions (Do Not Modify)

First, import the math library so that you can use predefined functions (such as math.pi and math.log).

In [1]:
# we put import functions at the beginning
import math  
from tabulate import tabulate 


The three functions below are the functions we will perform numerical integration on in this lab.  

In [2]:
# the example function integrated in calculus excerpt by Stein pp.370-372
# do NOT change this function
def f0(x):
    return 1.0/(1.0 + x*x)

In [3]:
# the function to integrate for Question 1
# do NOT change this function
def f1(x):
    return 1.0/math.sqrt(x*x-9.0)

In [4]:
# the function to integrate for Question 2
# do NOT change this function
def f2(x):
    return (1.0/math.sqrt(2*math.pi))*math.exp(-math.pow(x,2)/2)

We also need a means of computing the error for each of our estimates. The function below provides a standard format for printing out results.  You will use this function to report your results.

In [5]:
# Note: Do NOT change this function.
def print_result(n,estimate,answer):
    '''This function formats a standard print statement to use.
       n          is the number of subintervals
       estimate   is the estimate of the integral
       answer     is the correct answer
    '''
    error = math.fabs(estimate - answer)
    print("Subintervals: %4d,  Estimate: %f,  Error: %f" % (n,estimate,error))
# end of function print_result

In [6]:
# Let's see how this function works.
# suppose using 10 intervals we estimated a value of 99.5 when the real answer was 100.0
print_result(10,99.5,100.0)

Subintervals:   10,  Estimate: 99.500000,  Error: 0.500000


#### We will use this function to print out the results of our integral estimates below.

### Task 3 (Graded)

This file also contains three additional functions that you will be expected to modify:

* ```rectangular(f, a, b, n)```
* ```trapezoidal(f, a, b, n)``` 
* ```simpsons(f, a, b, n)```

These functions all use the same four parameters:

1. the name of the function to integrate, denoted f
2. the lower limit of integration, denoted a
3. the upper limit of integration, denoted b, and
4. the number of subintervals to be used in the estimate, denoted n.

Modify these functions to implement the algorithms specified in the handouts and test them. How will you validate that they are working properly?

#### Numerical Integration Functions (Modify These Functions in Place)

##### Rectangular Method

In [16]:
def rectangular(f, a, b, n):
    '''This function estimates an integral using the common "rectangular" method:
       f      is the function to integrate
       a, b   are the limits of integration
       n      is the number of subintervals
    '''
    h = (b-a)/n # this is my width 
    sum1 = 0.0
    i = 1
    while i <=  n: 
        x = a + i*h
        sum1 += f(x) #f(x) is my output  
        i += 1
    return round(h * sum1,4)  # your code goes here in place of this dummy statement
# end of rectangular

In [17]:
rectangular(f1, 4, 5, 1000)

0.3032

#### Trapezoidal Method

In [18]:
def trapezoidal(f, a, b, n):
    '''This function estimates an integral using the "trapezoidal" method:
       f      is the function to integrate
       a, b   are the limits of integration
       n      is the number of subintervals
    '''
    h = (b-a)/n
    sum2 = 0.0
    i = 0
    while i <= n:
        x = a + i*h
        if i == 0 or i ==n:
            sum2 += f(x) # Why is it not enough to just have 'f' as per the original argument
        else: 
            sum2 += 2*f(x)
        i += 1
    return round((h/2)*sum2,4) # your code goes here in place of this dummy statement

# end of trapezoidal

In [19]:
trapezoidal(f1,4,5,1000)

0.3032

#### Simpson's Method

In [20]:
def simpsons(f, a, b, n):
    '''This function estimates an integral using "simpsons" method:
       f      is the function to integrate
       a, b   are the limits of integration
       n      is the number of subintervals
    '''
    h = (b-a)/n
    sum3 = 0
    i = 0
    while i <= n:
        x = a + i*h
        if i == 0 or  i == n:
            sum3 += f(x)
        elif (i%2 == 0):
            sum3 += 2*f(x)
        else:
            sum3 += 4*f(x)
        i += 1
    return round((h/3) *sum3,4) # your code goes here in place of this dummy statement
    
# end of simpsons

In [21]:
simpsons(f1,4,5,25)

0.2999

### Task 4 (Graded)

Use the example provided in the Stein reference and the provided ```f0(x)``` to validate the performance of your numerical integration functions.  The two code blocks below provide the means for testing the performance of your functions:

In [22]:
q1_answer = math.pi/4

In [23]:
# note that we can run all our estimates 'at once'...
print_result(4,rectangular(f0,0.0,1.0,4),q1_answer)
print_result(4,trapezoidal(f0,0.0,1.0,4),q1_answer)
print_result(4,simpsons(f0,0.0,1.0,4),q1_answer)

Subintervals:    4,  Estimate: 0.720300,  Error: 0.065098
Subintervals:    4,  Estimate: 0.782800,  Error: 0.002598
Subintervals:    4,  Estimate: 0.785400,  Error: 0.000002


### Question 1:  How do you know (or why do you believe) that your functions are working correctly?  Double-click into the text block immediately below to provide an answer.

I know that my functions work correctly because i integraded the function by hand and got 0.7853981634

### Question 2:  . Estimate the integral  

$$\int_{4}^{5} \frac{1}{\sqrt{x^2 - 9}} dx$$  

### for n of 10, 25, 50, 100,  250, and 1000 using each of the three methods discussed in the lecture.  Use the code blocks below to record your results for each case.  For this integral, the correct answer is: 

$$ \ln \Bigg(  \frac{9}{4 + \sqrt{7}} \Bigg)$$

The four code blocks below (when executed) should answer Question \#2 **provided you have correctly implemented the algorithms above.**

In [24]:
q2_answer = math.log(9.0 / (4.0 + math.sqrt(7))) 

In [25]:
q2_answer

0.30324682744420395

In [26]:
# Question 2, rectangular method
# here's the function call for n=10
print_result(10,rectangular(f1,4.0,5.0,10),q2_answer)
print_result(25,rectangular(f1,4.0,5.0,25),q2_answer)
print_result(50,rectangular(f1,4.0,5.0,50),q2_answer)
print_result(100,rectangular(f1,4.0,5.0,100),q2_answer)
print_result(250,rectangular(f1,4.0,5.0,250),q2_answer)
print_result(1000,rectangular(f1,4.0,5.0,1000),q2_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.297000,  Error: 0.006247
Subintervals:   25,  Estimate: 0.300700,  Error: 0.002547
Subintervals:   50,  Estimate: 0.302000,  Error: 0.001247
Subintervals:  100,  Estimate: 0.302600,  Error: 0.000647
Subintervals:  250,  Estimate: 0.303000,  Error: 0.000247
Subintervals: 1000,  Estimate: 0.303200,  Error: 0.000047


In [27]:
1 / (0.002541 / 0.001275)

0.5017709563164109

In [28]:
# Question 2, trapezoidal method
# here's the function call for n=10
print_result(10,trapezoidal(f1,4.0,5.0,10),q2_answer)
print_result(25,trapezoidal(f1,4.0,5.0,25),q2_answer)
print_result(50,trapezoidal(f1,4.0,5.0,50),q2_answer)
print_result(100,trapezoidal(f1,4.0,5.0,100),q2_answer)
print_result(250,trapezoidal(f1,4.0,5.0,250),q2_answer)
print_result(1000,trapezoidal(f1,4.0,5.0,1000),q2_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.303400,  Error: 0.000153
Subintervals:   25,  Estimate: 0.303300,  Error: 0.000053
Subintervals:   50,  Estimate: 0.303300,  Error: 0.000053
Subintervals:  100,  Estimate: 0.303200,  Error: 0.000047
Subintervals:  250,  Estimate: 0.303200,  Error: 0.000047
Subintervals: 1000,  Estimate: 0.303200,  Error: 0.000047


In [29]:
(0.000005/0.000018)

0.2777777777777778

In [30]:
# Question 2, Simpson's method
# here's the function call for n=10
print_result(10,simpsons(f1,4.0,5.0,10),q2_answer)
print_result(25,simpsons(f1,4.0,5.0,25),q2_answer)
print_result(50,simpsons(f1,4.0,5.0,50),q2_answer)
print_result(100,simpsons(f1,4.0,5.0,100),q2_answer)
print_result(250,simpsons(f1,4.0,5.0,250),q2_answer)
print_result(1000,simpsons(f1,4.0,5.0,1000),q2_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.303200,  Error: 0.000047
Subintervals:   25,  Estimate: 0.299900,  Error: 0.003347
Subintervals:   50,  Estimate: 0.303200,  Error: 0.000047
Subintervals:  100,  Estimate: 0.303200,  Error: 0.000047
Subintervals:  250,  Estimate: 0.303200,  Error: 0.000047
Subintervals: 1000,  Estimate: 0.303200,  Error: 0.000047


### Question 3:  Estimate the integral

$$ \int_{0}^{2.5} \frac{1}{\sqrt{2 \pi}} e^{-z^2/2} dz$$

### for the following values of n for n of 10, 25, 50, 100,  250, and 1000  using each of the three methods discussed in lecture.  For each case, also record the absolute errors in your computation. For this integral, the correct answer is: 0.493790335.

The four code blocks below (when executed) should answer Question \#3 **provided you have correctly implemented the algorithms above.**

In [34]:
q3_answer = 0.493790335  # taken from lookup table - see solutions handout

In [35]:
# Question 3, rectangular method
# here's the function call for n=10
print_result(10,rectangular(f2,0,2.5,10),q3_answer)
print_result(25,rectangular(f2,0,2.5,25),q3_answer)
print_result(50,rectangular(f2,0,2.5,50),q3_answer)
print_result(100,rectangular(f2,0,2.5,100),q3_answer)
print_result(250,rectangular(f2,0,2.5,250),q3_answer)
print_result(1000,rectangular(f2,0,2.5,1000),q3_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.445900,  Error: 0.047890
Subintervals:   25,  Estimate: 0.474700,  Error: 0.019090
Subintervals:   50,  Estimate: 0.484200,  Error: 0.009590
Subintervals:  100,  Estimate: 0.489000,  Error: 0.004790
Subintervals:  250,  Estimate: 0.491900,  Error: 0.001890
Subintervals: 1000,  Estimate: 0.493300,  Error: 0.000490


In [36]:
# Question 3, trapezoidal method
# here's the function call for n=10
print_result(10,trapezoidal(f2,0,2.5,10),q3_answer)
print_result(25,trapezoidal(f2,0,2.5,25),q3_answer)
print_result(50,trapezoidal(f2,0,2.5,50),q3_answer)
print_result(100,trapezoidal(f2,0,2.5,100),q3_answer)
print_result(250,trapezoidal(f2,0,2.5,250),q3_answer)
print_result(1000,trapezoidal(f2,0,2.5,1000),q3_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.493600,  Error: 0.000190
Subintervals:   25,  Estimate: 0.493800,  Error: 0.000010
Subintervals:   50,  Estimate: 0.493800,  Error: 0.000010
Subintervals:  100,  Estimate: 0.493800,  Error: 0.000010
Subintervals:  250,  Estimate: 0.493800,  Error: 0.000010
Subintervals: 1000,  Estimate: 0.493800,  Error: 0.000010


In [37]:
# Question 3, Simpson's method
# here's the function call for n=10
print_result(10,simpsons(f2,0,2.5,10),q3_answer)
print_result(25,simpsons(f2,0,2.5,25),q3_answer)
print_result(50,simpsons(f2,0,2.5,50),q3_answer)
print_result(100,simpsons(f2,0,2.5,100),q3_answer)
print_result(250,simpsons(f2,0,2.5,250),q3_answer)
print_result(1000,simpsons(f2,0,2.5,1000),q3_answer)
# include here additional function calls (as appropriate) for n=25, 50, 100, 250, 1000


Subintervals:   10,  Estimate: 0.493800,  Error: 0.000010
Subintervals:   25,  Estimate: 0.493100,  Error: 0.000690
Subintervals:   50,  Estimate: 0.493800,  Error: 0.000010
Subintervals:  100,  Estimate: 0.493800,  Error: 0.000010
Subintervals:  250,  Estimate: 0.493800,  Error: 0.000010
Subintervals: 1000,  Estimate: 0.493800,  Error: 0.000010


### Question 4: Based on your results in Question 1 and Question 2 above, complete the following three statements (you may add some code blocks below to figure out the answer to this question as needed).

Note: double-click into cells to replace "YOUR ANSWER HERE".

* For the Rectangular Method, cutting the interval h in half results in cutting the error by: 

Cutting the interval h in half results in cutting the error by 50%


* For the Trapezoidal Method, cutting the interval h in half results in cutting the error by:

 Cutting the interval h in half results in cutting the error by 27.78%
 

* For the Simpson's Method, cutting the interval h in half results in cutting the error by:

Do not have enough data to draw a conclusion 


### Task 5 (Graded)

Write another function to evaluate the integral 

$$ \int_0^1 \int_0^1 ( x^2 + y) \ dy \ dx $$

via a two-dimensional version of the rectangular method (evaluate the function at the right-side of each subinterval). In this problem, you will need to iterate over both the x and y dimensions. Specifically, you should write a function called 

**rect2d(ax, bx, nx, ay, by, ny)** 

where ```[ax, bx]``` denotes the range of integration for the x variable and ```nx``` denotes the number of intervals in this dimension.  Similarly, ```[ay, by]``` denotes the range of integration for the y variable, and ```ny``` denotes the number of intervals in this dimension. To compute this integral, you will need to break the range for each dimension into ```nx``` and ```ny``` subintervals, respectively. 

Note: you do not need to pass an argument indicating which function to integrate — *your code needs only to work for this single function.*

Use the code block below to develop your function:

In [38]:
def f(x,y):
    return (x*x + y)

In [39]:
#estimates 2d integral using rectangular method
def rect2d(ax, bx, nx, ay, by, ny):
    hx = (bx-ax)/nx
    hy = (by-ay)/ny
    sum4 = 0
    i = 1
    while i <= nx:
        x = ax + i*hx
        j = 1 
        while  j <= ny:
            y = ay + j*hy
            sum4 += f(x,y)
            j += 1
        i += 1 
    return hx*hy*sum4 # your code goes here in place of this dummy statement
# end of rect2d  

In [40]:
rect2d(0,1,20,0,1,20) #testing the code above 

0.8837500000000004

### Question 5:  Use the Rectangular Method to estimate the integral 

$$\int_0^1 \int_0^1 ( x^2 + y) \ dy \ dx $$

###     for the following values of nx and ny:  

| $n_x$	| $n_y$	|
|:-:	|---	|
|    4	|    4	|
|   16	|   16	|
|   64 	|   64 	|
|  256 	|  256 	|
| 1024 	| 1024 	|

The code block below (when executed) should answer Question \#5 **provided you have correctly implemented the algorithm in Task 5 above.**

In [41]:
q5_answer = 5/6
float(5/6) # needed to print the number to figure 

0.8333333333333334

In [42]:
# okay, here are all the function calls for this one...
print_result(4,rect2d(0.0,1.0,4,0.0,1.0,4),q5_answer)
print_result(16,rect2d(0.0,1.0,16,0.0,1.0,16),q5_answer)
print_result(64,rect2d(0.0,1.0,64,0.0,1.0,64),q5_answer)
print_result(256,rect2d(0.0,1.0,256,0.0,1.0,256),q5_answer)
print_result(10,rect2d(0.0,1.0,1024,0.0,1.0,1024),q5_answer)

Subintervals:    4,  Estimate: 1.093750,  Error: 0.260417
Subintervals:   16,  Estimate: 0.896484,  Error: 0.063151
Subintervals:   64,  Estimate: 0.848999,  Error: 0.015666
Subintervals:  256,  Estimate: 0.837242,  Error: 0.003909
Subintervals:   10,  Estimate: 0.834310,  Error: 0.000977
